File: auth-service/src/main/java/com/eirsteir/coffeewithme/authservice/security/JwtUsernameAndPasswordAuthenticationFilter.java
package com.eirsteir.coffeewithme.authservice.security;

import com.eirsteir.coffeewithme.authservice.domain.Account;
import com.eirsteir.coffeewithme.commons.security.JwtConfig;
import com.eirsteir.coffeewithme.commons.security.JwtUtils;
import com.eirsteir.coffeewithme.commons.security.UserDetailsImpl;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.util.Collections;
import javax.servlet.FilterChain;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
import org.springframework.web.cors.CorsUtils;

@Slf4j
public class JwtUsernameAndPasswordAuthenticationFilter
    extends UsernamePasswordAuthenticationFilter {

  private AuthenticationManager authManager;

  private final JwtConfig jwtConfig;

  public JwtUsernameAndPasswordAuthenticationFilter(
      AuthenticationManager authManager, JwtConfig jwtConfig) {
    this.authManager = authManager;
    this.jwtConfig = jwtConfig;

    this.setRequiresAuthenticationRequestMatcher(
        new AntPathRequestMatcher(jwtConfig.getUri() + "/login", "POST"));
  }

  @Override
  public Authentication attemptAuthentication(
      HttpServletRequest request, HttpServletResponse response) throws AuthenticationException {

    if (CorsUtils.isPreFlightRequest(request)) {
      response.setStatus(HttpServletResponse.SC_OK);
      return null;
    }

    try {
      Account account = new ObjectMapper().readValue(request.getInputStream(), Account.class);
      log.debug("[x] Found credentials: {}", account);

      UsernamePasswordAuthenticationToken authToken =
          new UsernamePasswordAuthenticationToken(
              account.getEmail(), account.getPassword(), Collections.emptyList());

      log.debug("[x] Attempting to authenticate token: {}", authToken);
      return authManager.authenticate(authToken);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  protected void successfulAuthentication(
      HttpServletRequest request,
      HttpServletResponse response,
      FilterChain chain,
      Authentication auth)
      throws IOException {

    log.debug("[x] Authentication principal: {}", auth.getPrincipal());

    UserDetailsImpl principal = (UserDetailsImpl) auth.getPrincipal();

    String token = JwtUtils.createJwtToken(jwtConfig, auth, principal);

    response.addHeader(jwtConfig.getHeader(), jwtConfig.getPrefix() + token);
    response.setContentType("application/json");
    response.setCharacterEncoding("UTF-8");
    response.getWriter().write(token);
  }
}

File: auth-service/src/main/kotlin/com/eirsteir/coffeewithme/authservice/security/JwtUsernameAndPasswordAuthenticationFilter.kt
package com.eirsteir.coffeewithme.authservice.security

import com.eirsteir.coffeewithme.authservice.domain.Account
import com.eirsteir.coffeewithme.commons.security.JwtConfig
import com.eirsteir.coffeewithme.commons.security.JwtUtils
import com.eirsteir.coffeewithme.commons.security.UserDetailsImpl
import com.fasterxml.jackson.databind.ObjectMapper
import lombok.extern.slf4j.Slf4j
import mu.KotlinLogging
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.security.web.util.matcher.AntPathRequestMatcher
import org.springframework.web.cors.CorsUtils
import java.io.IOException
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse


private val logger = KotlinLogging.logger {}


class JwtUsernameAndPasswordAuthenticationFilter(
    private val authManager: AuthenticationManager,
    private val jwtConfig: JwtConfig
) : UsernamePasswordAuthenticationFilter() {

    override fun attemptAuthentication(
        request: HttpServletRequest, response: HttpServletResponse
    ): Authentication? {
        if (CorsUtils.isPreFlightRequest(request)) {
            response.status = HttpServletResponse.SC_OK
            return null
        }

        return try {
            val account = ObjectMapper().readValue(request.inputStream, Account::class.java)
            logger.debug("[x] Found credentials: $account")

            val authToken = UsernamePasswordAuthenticationToken(
                account.email, account.password, emptyList()
            )
            logger.debug("[x] Attempting to authenticate token: $authToken")
            authManager.authenticate(authToken)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    override fun successfulAuthentication(
        request: HttpServletRequest,
        response: HttpServletResponse,
        chain: FilterChain,
        auth: Authentication
    ) {
        logger.debug("[x] Authentication principal: ${auth.principal}")

        val principal = auth.principal as UserDetailsImpl
        val token = JwtUtils.createJwtToken(jwtConfig, auth, principal)

        response.addHeader(jwtConfig.header, jwtConfig.prefix + token)
        response.contentType = "application/json"
        response.characterEncoding = "UTF-8"
        response.writer.write(token)
    }

    init {
        setRequiresAuthenticationRequestMatcher(
            AntPathRequestMatcher(jwtConfig.uri + "/login", "POST")
        )
    }
}